File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices or len(prices) == 1:
            return 0
        
        # Initialize variables
        best_profit = 0
        best_day_to_buy = 0

        # Iterate through the prices
        for i in range(1, len(prices)):
            if prices[i] < prices[best_day_to_buy]:
                best_day_to_buy = i
            else:
                # Calculate and update the best profit if selling on the current day
                best_profit = max(best_profit, prices[i] - prices[best_day_to_buy])

        return best_profit
                
                
                
            

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cs
public class Solution {
    public int MaxProfit(int[] prices) {
    
    int max_profit = 0;
    int min_prices = prices[0];

    for(int i=1; i<prices.Length; i++){
        // If we've a better price to buy(lower)
        if(prices[i] < min_prices ){
            min_prices = prices[i];
        }
        if(prices[i] - min_prices > max_profit){
            max_profit = prices[i] - min_prices;
        }
    }
    return max_profit;


        // int P = prices.Length;
        // int profit = 0;
        // for(int i=0; i<P;i++){
        //     for(int j=i+1;j<P;j++){
        //         if(profit < prices[j] - prices[i]){
        //             profit = prices[j] - prices[i];
        //         }
        //     } 
        // }
        // return profit; 
    }
}

    //     if(P==1){
    //         return 0;
    //     }

    //     int dayToBuy = 0;
    //     for(int i=1; i<P-1; i++){
    //         if(prices[dayToBuy] > prices[i]){
    //             dayToBuy = i;
    //         }
    //     }
    //     int dayToSell = dayToBuy+1;
    //     for(int i=dayToBuy+2;i<P;i++){
    //         if(prices[dayToSell] < prices[i]){
    //             dayToSell = i;
    //         }
    //     }

    //     if(prices[dayToBuy] > prices[dayToSell]){
    //         return 0;
    //     }
    //     return prices[dayToSell] - prices[dayToBuy];
    // }
